pyro/infer/tracegraph_elbo.py

```diff
@@ -147,24 +164,48 @@ def _compute_elbo_non_reparam(guide_trace, guide_vec_md_nodes,  #
             avg_downstream_cost_old.data = avg_downstream_cost_new.data  # XXX copy_() ?
             baseline += avg_downstream_cost_old
         if use_nn_baseline:
-            # block nn_baseline_input gradients except in baseline loss
             baseline += nn_baseline(detach_iterable(nn_baseline_input))
         elif use_baseline_value:
             # it's on the user to make sure baseline_value tape only points to baseline params
+            # if using baseline_value
             baseline += baseline_value
         if use_nn_baseline or use_baseline_value:
             # accumulate baseline loss
-            baseline_loss += torch.pow(downstream_cost.detach() - baseline, 2.0).sum()
-
+            if not use_lax:
+                baseline_loss += ((downstream_cost.detach() - baseline)).norm()
         guide_log_pdf = guide_site[log_pdf_key] / guide_site["scale"]  # not scaled by subsampling
         if use_nn_baseline or use_decaying_avg_baseline or use_baseline_value:
             if downstream_cost.size() != baseline.size():
                 raise ValueError("Expected baseline at site {} to be {} instead got {}".format(
                     node, downstream_cost.size(), baseline.size()))
-            downstream_cost = downstream_cost - baseline
-        surrogate_elbo += (guide_log_pdf * downstream_cost.detach()).sum()
+            downstream_cost = downstream_cost.detach() - baseline
+        # compute model/guide gradients (these still need to be scaled by cost terms)
+        grads_log_q = autograd.grad(guide_log_pdf, trainable_params_minus_bl_params,
+                                    create_graph=True, allow_unused=True)
+        for k in range(len(grads_log_q)):
+            g, p = grads_log_q[k], trainable_params_minus_bl_params[k]
+            # assemble the reinforce-like term
+            cost_times_grad = downstream_cost * g.detach() if g is not None else None
+            if cost_times_grad is None:
+                continue  # log_q didn't depend on parameter p
+            if p.grad is None:
+                p.grad = -cost_times_grad  # this is the first time we've differentiated w.r.t. p
+            else:
+                p.grad -= cost_times_grad  # otherwise accumulate p gradient
+            if use_lax:  # if using lax, we keep track of the gradients to compute the lax baseline_loss
+                assert(use_nn_baseline), "lax estimator must use nn_baseline and not baseline_value"
+                if full_gradients[k] is None:
+                    full_gradients[k] = cost_times_grad
+                else:
+                    full_gradients[k] += cost_times_grad
+
+    # if any nodes have use_lax then construct lax baseline_loss
```

Comment:
Punctuate your comments. The lack of period in this comment is confusing and burdens readers with an unnecessary inference task of parsing unstructured text.